File: apple/Sources/LogSense/Internal/Search/TextMatcher.swift
import Foundation

/// Find-bar options: match case / whole word / regex.
internal struct SearchQuery: Equatable {
    var text: String = ""
    var matchCase: Bool = false
    var wholeWord: Bool = false
    var regex: Bool = false

    var isActive: Bool { !text.isEmpty }
}

/// Compiles a `SearchQuery` into a reusable matcher — literal / match-case / whole-word / regex.
/// An invalid regex (or half-typed pattern) resolves to "matches nothing" instead of throwing.
internal struct TextMatcher {
    private let regex: NSRegularExpression?
    private let literal: String?
    private let ignoreCase: Bool
    private let matchNothing: Bool

    /// True when `input` contains at least one match. Empty query matches everything.
    func matches(_ input: String) -> Bool {
        if matchNothing { return false }
        if let regex {
            return regex.firstMatch(in: input, range: NSRange(input.startIndex..., in: input)) != nil
        }
        if let literal {
            return input.range(of: literal, options: ignoreCase ? [.caseInsensitive] : []) != nil
        }
        return true
    }

    /// Non-empty match ranges within `input` as UTF-16 offsets, for highlighting.
    /// Empty when the query is blank.
    func ranges(_ input: String) -> [Range<Int>] {
        if matchNothing { return [] }
        if let regex {
            return regex.matches(in: input, range: NSRange(input.startIndex..., in: input))
                .map(\.range)
                .filter { $0.length > 0 }
                .map { $0.location..<($0.location + $0.length) }
        }
        guard let literal, !literal.isEmpty else { return [] }
        var out: [Range<Int>] = []
        let haystack = input as NSString
        var search = NSRange(location: 0, length: haystack.length)
        let options: NSString.CompareOptions = ignoreCase ? [.caseInsensitive, .literal] : [.literal]
        while search.length > 0 {
            let found = haystack.range(of: literal, options: options, range: search)
            if found.location == NSNotFound { break }
            out.append(found.location..<(found.location + found.length))
            let next = found.location + found.length
            search = NSRange(location: next, length: haystack.length - next)
        }
        return out
    }

    static func from(_ query: SearchQuery) -> TextMatcher {
        let ignoreCase = !query.matchCase
        if query.text.isEmpty {
            return TextMatcher(regex: nil, literal: nil, ignoreCase: ignoreCase, matchNothing: false)
        }
        let options: NSRegularExpression.Options = ignoreCase ? [.caseInsensitive] : []
        if query.regex {
            return compile(query.text, options: options, ignoreCase: ignoreCase)
        }
        if query.wholeWord {
            let escaped = NSRegularExpression.escapedPattern(for: query.text)
            return compile("\\b\(escaped)\\b", options: options, ignoreCase: ignoreCase)
        }
        return TextMatcher(regex: nil, literal: query.text, ignoreCase: ignoreCase, matchNothing: false)
    }

    private static func compile(
        _ pattern: String, options: NSRegularExpression.Options, ignoreCase: Bool
    ) -> TextMatcher {
        if let regex = try? NSRegularExpression(pattern: pattern, options: options) {
            return TextMatcher(regex: regex, literal: nil, ignoreCase: ignoreCase, matchNothing: false)
        }
        return TextMatcher(regex: nil, literal: nil, ignoreCase: ignoreCase, matchNothing: true)
    }
}

File: android/logsense/src/main/java/com/msabhi/logsense/internal/search/TextMatcher.kt
package com.msabhi.logsense.internal.search

/** Android-Studio-style find options for the search bar. */
internal data class SearchQuery(
    val text: String = "",
    val matchCase: Boolean = false,
    val wholeWord: Boolean = false,
    val regex: Boolean = false,
) {
    val isActive: Boolean get() = text.isNotEmpty()
}

/**
 * Compiles a [SearchQuery] into a reusable matcher — literal / match-case / whole-word / regex,
 * mirroring Android Studio's find bar. An invalid regex (or half-typed pattern) resolves to
 * "matches nothing" instead of throwing.
 */
internal class TextMatcher private constructor(
    private val regex: Regex?,
    private val literal: String?,
    private val ignoreCase: Boolean,
    private val matchNothing: Boolean,
) {
    /** True when [input] contains at least one match. Empty query matches everything. */
    fun matches(input: String): Boolean = when {
        matchNothing -> false
        regex != null -> regex.containsMatchIn(input)
        literal != null -> input.contains(literal, ignoreCase)
        else -> true
    }

    /** Non-empty match ranges within [input], for highlighting. Empty when the query is blank. */
    fun ranges(input: String): List<IntRange> = when {
        matchNothing -> emptyList()
        regex != null -> regex.findAll(input).map { it.range }.filter { it.first <= it.last }.toList()
        literal != null && literal.isNotEmpty() -> buildList {
            var i = input.indexOf(literal, 0, ignoreCase)
            while (i >= 0) {
                add(i until i + literal.length)
                i = input.indexOf(literal, i + literal.length, ignoreCase)
            }
        }
        else -> emptyList()
    }

    companion object {
        fun from(query: SearchQuery): TextMatcher {
            val ignoreCase = !query.matchCase
            if (query.text.isEmpty()) return TextMatcher(null, null, ignoreCase, matchNothing = false)
            val opts = if (ignoreCase) setOf(RegexOption.IGNORE_CASE) else emptySet()
            return when {
                query.regex -> compile(query.text, opts, ignoreCase)
                query.wholeWord -> compile("""\b${Regex.escape(query.text)}\b""", opts, ignoreCase)
                else -> TextMatcher(null, query.text, ignoreCase, matchNothing = false)
            }
        }

        private fun compile(pattern: String, opts: Set<RegexOption>, ignoreCase: Boolean): TextMatcher =
            runCatching { Regex(pattern, opts) }
                .map { TextMatcher(it, null, ignoreCase, matchNothing = false) }
                .getOrElse { TextMatcher(null, null, ignoreCase, matchNothing = true) }
    }
}
